### PR TITLE
[binary addons] automatically set the proper platform specific library in addon.xml

### PR DESCRIPTION
**Pull request by Montellese**
The main goal of this PR is for the cmake buildsystem to be able to automatically fill in the platform and its specific library name into the `addon.xml`. This is a necessary step if we ever want to be able to provide binary addons through platform specific repositories because then we can't ship an addon with an `addon.xml` that contains `library_foo` definitions for platforms that aren't supported by the addon.

The main functionality for this has been added with 825dcc2. For the new functionality to be used the addon needs to provide an `addon.xml.in` instead of an `addon.xml` with the following line: `library_@PLATFORM@="@LIBRARY_FILENAME@"` (see e.g. https://github.com/Montellese/audioencoder.flac/blob/dynamic_platform/audioencoder.flac/addon.xml.in#L13). Binary addons with `addon.xml` instead of `addon.xml.in` are still supported and remain untouched/unmodified.

When I looked at this I noticed some inconcsistencies in our addon implementation which are fixed by the other commits:
- There was a duplicate implementation of parsing the `library_<platform>="<library name>"` attribute from `addon.xml`. I've moved it to a helper method in `CAddonMgr`.
- Some binary addons use the `library_freebsd` attribute in `addon.xml` but it's not supported by core.
- For easier implementation of the main feature I extended the possible attributes by adding support for `library_rbpi`, `library_ios`, `library_darwin` and `library_windows` because those are the target platform names we use in the binary addon buildsystem. I only tested `library_windows` so would be nice if devs for the respective platforms could check my `#ifdef`ery.

The last commit is only there for testing and illustration and should be dropped before merging.

### CODE REVIEW

**Comment by Montellese** [comment]
Updated:
* also dynamically fill the `<platform>` tag in `addon.xml.in` (see audioencoder.flac example)
* rename `darwin` to `osx` in `CORE_SYSTEM_NAME` as requested by @Memphiz. This requires a small change to `platform` and `kodi-platform`.

File: project/cmake/scripts/common/addon-helpers.cmake [hunk 1]
```diff
@@ -34,6 +39,17 @@ macro (build_addon target prefix libs)
     SET_TARGET_PROPERTIES(${target} PROPERTIES PREFIX "lib")
   ENDIF(OS STREQUAL "android")
 
+  GET_TARGET_PROPERTY(LIBRARY_LOCATION ${target} LOCATION)
```
Comment (wsnipex) [inline project/cmake/scripts/common/addon-helpers.cmake]: this has been deprecated by http://www.cmake.org/cmake/help/v3.0/policy/CMP0026.html
see https://github.com/wsnipex/xbmc/commit/4284ce9c4675dd52354c7e1e5f50c66fe9a0a4df for a POC that works around it on linux. Something similar would be needed for the VS project file.

Comment (Montellese) [inline project/cmake/scripts/common/addon-helpers.cmake]: I know and I tried `$<TARGET_FILE:${target}>` but it always got interpreted literally with no replacement of the generator.

**Comment by Montellese** [comment]
@wsnipex: Thanks for the hint on how to get the `$<TARGET_FILE>` and `$TARGET_FILE_NAME>` generators working. I've incorporated your approach into this PR and it makes it much cleaner.

File: project/cmake/scripts/common/addon-helpers.cmake [hunk 2]
```diff
@@ -34,6 +39,20 @@ macro (build_addon target prefix libs)
     SET_TARGET_PROPERTIES(${target} PROPERTIES PREFIX "lib")
   ENDIF(OS STREQUAL "android")
 
+  SET(LIBRARY_LOCATION $<TARGET_FILE:${target}>)
+  SET(LIBRARY_FILENAME $<TARGET_FILE_NAME:${target}>)
+
+  # if there's an addon.xml.in we need to generate the addon.xml
+  SET(GENERATED_ADDONXML OFF)
```
Comment (wsnipex) [inline project/cmake/scripts/common/addon-helpers.cmake]: This is not needed anymore

**Comment by wsnipex** [comment]
Very nice!
Since addon.XML now always contains only one correct lib, we dont need the ifdefs in getplatform anymore, vor am I missing something?

**Comment by Montellese** [comment]
I'd keep the platform-specific library/platform parsing for two reasons:
* backwards compatibility for addons that haven't been moved to the new approach yet
* users can still download some random zip from somewhere and try to install it in Kodi and load the addon. in that case refusing to install/load the addon due to mismatching platforms is IMO nicer.

**Comment by Montellese** [comment]
@Memphiz: could you take a look at the OSX related changes?

**Comment by wsnipex** [comment]
Yeah agreed, I missed backwards compatibility. But we can deprecate the old Methode and drop the parsing in v17 vor so

**Comment by Memphiz** [comment]
@Montellese looks fine

**Comment by Montellese** [comment]
How should we proceed? We need to
* merge https://github.com/Pulse-Eight/platform/pull/14 into platform
* create a new archive that can be downloaded for our binary addons depends system
* merge https://github.com/xbmc/kodi-platform/pull/11
* bump the revision of `kodi-platform` in our binary addons depends system
* merge this PR
* adjust all binary addons to use the new `addons.xml.in`

**Comment by hudokkow** [comment]
Sounds like a plan and count me in to do monkey work if needed but AFAIK @opdenkamp is busy renovating is new home.

**Comment by Montellese** [comment]
@wsnipex: Are you ok with that procedure? If so could you merge https://github.com/Pulse-Eight/platform/pull/14 (as I don't have write access)?

**Comment by wsnipex** [comment]
sure, platform change merged.

**Comment by Montellese** [comment]
Thanks. Will create the source package for it and update this PR.

**Comment by Montellese** [comment]
I just realized that we are downloading a `.tar.gz` for the `platform` library so we'll need a version bump from `1.0.10` to `1.0.11` to be able to update the dependency package on our mirrors :-/

**Comment by opdenkamp** [comment]
On 05-08-15 07:52, Sascha Montellese wrote:
> (as I don't have write
> access)?

you never asked for write access ;) you've received an invite

**Comment by Montellese** [comment]
@opdenkamp: Thanks, much appreciated. I never needed it until now ;-)

**Comment by Montellese** [comment]
jenkins build this please

**Comment by Montellese** [comment]
Two addons failed to build because of mirror hickups when downloading some dependencies.

**Comment by hudokkow** [comment]
@Montellese, there's something wrong with either this PR or with my add-on PRs -> http://pastebin.com/raw.php?i=Ld5exP1K

**Comment by Montellese** [comment]
@hudokkow: It might be related to the `;` in the `addon.xml.in` which is used by cmake as a delimiter for lists of strings. I'll have to take a closer look later because @wsnipex wrote that part of the code.

File: project/cmake/scripts/common/addon-helpers.cmake [hunk 3]
```diff
@@ -40,6 +45,18 @@ macro (build_addon target prefix libs)
...
 
+  SET(LIBRARY_LOCATION $<TARGET_FILE:${target}>)
+  SET(LIBRARY_FILENAME $<TARGET_FILE_NAME:${target}>)
+
+  # if there's an addon.xml.in we need to generate the addon.xml
+  IF(EXISTS ${PROJECT_SOURCE_DIR}/${target}/addon.xml.in)
+    SET(PLATFORM ${CORE_SYSTEM_NAME})
+
+    FILE(READ ${PROJECT_SOURCE_DIR}/${target}/addon.xml.in addon_file)
+    STRING(CONFIGURE ${addon_file} addon_file_conf @ONLY)
```
Comment (Montellese) [inline project/cmake/scripts/common/addon-helpers.cmake]: We most likely need to put double quotes around `${addon_file}` and `${addon_file_conf}` to tell CMake that these are strings and not lists (delimited by semicolons).

**Comment by hudokkow** [comment]
hmm, ok. Thanks.

**Comment by Montellese** [comment]
@wsnipex: I've noticed a problem. The CMake generators you've introduced have been introduced with CMake 3.0 but we state
```
cmake_minimum_required(VERSION 2.8)
```
(in some places even `2.6` IIRC) and the version of cmake that we build with depends (in case that's relevant) is `2.8.X` as well. So either we need to go back to the old `get_target_property()` approach with `LOCATION` and live with the warnings on CMake 3.0+ or we need to update the minimum required version to `3.0`.

**Comment by Memphiz** [comment]
in the latter case the cmake in the native depends needs to be bumped aswell (so that we have that tool available for building on osx/ios slave)

**Comment by wsnipex** [comment]
hmm? http://www.cmake.org/cmake/help/v2.8.8/cmake.html
search for generator expression

**Comment by Montellese** [comment]
@wsnipex: I noticed this when I tried to setup my laptop and I installed CMake 2.8.2 and there I get errors that `$<TARGET_FILE_NAME:audioencoder.wav>` is not an installable file because it can't resolve the generator expression. Installing latest CMake solved the problem.

**Comment by wsnipex** [comment]
then I wonder how this ever built on jenkins. If it really is broken, we have to upgrade cmake. We'll need it sooner or later anyway.

**Comment by Montellese** [comment]
OMG this is ridiculous. I tried a few CMake versions on win32:
* 2.8.0 fails to extract ZIP files
* 2.8.2 fails to find the `git` executable
* 2.8.4, 2.8.6 and 2.8.8 fail because the integrated libcurl has been built without SSL support so we can't download from HTTPS (i.e. github)
* 2.8.10 and 2.8.12.2 (the last 2.8 release) can't resolve the generator expressions
* 3.0.2, 3.1.3, 3.2.3 and 3.3.0 (latest) seem to fail to properly put the cmake executable into the PATH variable but work fine otherwise

So at least on win32 cmake 2.8.X is a no-go. I don't know about any other platform though.

**Comment by metaron-uk** [comment]
@montellese: Is this change supposed to automatically download and use libcrossguid for you, or should you install it on your system somehow?

Since it was merged: ```git clean -xfd ; git reset --hard ; ./bootstrap; ./configure``` fails at:
```
checking for main in -ltinyxml... yes
checking for UUID... yes
checking "for libcrossguid"... configure: error: crossguid not found
```
I cloned libcrossguid but it doesn't seem to have a make / make install function?
(I'm on gentoo x86)

**Comment by hudokkow** [comment]
hudo@Toshiba:~/dev/kodiws$ apt-cache policy **libcrossguid-dev**
libcrossguid-dev:
  Installed: 0.1~git20150807.8f399e8~vivid
  Candidate: 0.1~git20150807.8f399e8~vivid
  Version table:
 *** 0.1~git20150807.8f399e8~vivid 0
        500 http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu/ vivid/main amd64 Packages
        500 http://ppa.launchpad.net/team-xbmc/xbmc-ppa-build-depends/ubuntu/ vivid/main amd64 Packages
        100 /var/lib/dpkg/status

**Comment by wsnipex** [comment]
crossguid has nothing to do with this PR.
see https://github.com/xbmc/xbmc/blob/master/tools/depends/target/crossguid/Makefile

**Comment by metaron-uk** [comment]
thanks @wsnipex / @hudokkow , sorry for the noise.

**Comment by Jalle19** [comment]
Could someone who knows how this new `addon.xml.in` file is supposed to work comment in http://forum.kodi.tv/showthread.php?tid=219166&pid=2083241#pid2083241 ?

**Pull request merged**